src/main/java/org/folio/dew/repository/AbstractFilesStorage.java

```diff
@@ -0,0 +1,332 @@
+package org.folio.dew.repository;
+
+import org.apache.commons.lang3.StringUtils;
+import org.folio.dew.config.properties.MinioClientProperties;
+import org.folio.s3.client.FolioS3Client;
+import org.folio.s3.client.PutObjectAdditionalOptions;
+import org.folio.s3.client.RemoteStorageWriter;
+import org.folio.s3.client.S3ClientFactory;
+import org.folio.s3.exception.S3ClientException;
+
+import java.io.BufferedReader;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.SequenceInputStream;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+import io.minio.http.Method;
+import lombok.extern.log4j.Log4j2;
+
+import static org.folio.dew.utils.Constants.PATH_SEPARATOR;
+import static org.folio.dew.utils.Constants.TMP_PATH_PREFIX;
+
+/** Custom extension of FolioS3Client with additional utility methods */
+@Log4j2
+public abstract class AbstractFilesStorage implements FolioS3Client {
+
+  private final FolioS3Client client;
+
+  private final MinioClientProperties properties;
+
+  protected AbstractFilesStorage(MinioClientProperties properties) {
+    this.client = S3ClientFactory.getS3Client(properties.toS3ClientProperties());
+    this.properties = properties;
+
+    this.createBucketIfNotExists();
+  }
+
+  @Override
+  public String upload(String path, String filename) {
+    return client.upload(sanitizePath(path), filename);
+  }
+
+  @Override
+  public String append(String path, InputStream newData) {
```

Comment:
Hi @kjain110 , the append() method should not be used at all, as it has been deprecated in the library. The current implementation first downloads all the data, updates it, and then re-uploads everything, which is expected to be very slow. The main idea is to refactor the logic to completely eliminate the use of the append() method, since it would require different implementations for AWS S3 and Minio in both cases.